File: SL-Back-end/app/schemas/market_quote.py
"""
시세 페이지 API 스키마
- 전체 종목 리스트 조회
- 정렬 기능 (거래량, 등락률, 거래대금, 시가총액)
"""

from typing import List, Optional
from pydantic import BaseModel, Field, ConfigDict
from enum import Enum


class SortBy(str, Enum):
    """정렬 기준"""
    VOLUME = "volume"  # 거래량 순
    CHANGE_RATE = "change_rate"  # 등락률 순
    TRADING_VALUE = "trading_value"  # 거래대금 순
    MARKET_CAP = "market_cap"  # 시가총액 순
    NAME = "name"  # 종목명 순


class SortOrder(str, Enum):
    """정렬 순서"""
    ASC = "asc"  # 오름차순
    DESC = "desc"  # 내림차순


class MarketQuoteItem(BaseModel):
    """시세 항목"""
    model_config = ConfigDict(populate_by_name=True)

    rank: int = Field(..., description="순위 (정렬 기준에 따라 변동)")
    name: str = Field(..., description="종목명")
    code: str = Field(..., description="종목 코드")
    price: int = Field(..., description="현재가 (종가)")
    change_amount: int = Field(..., serialization_alias="changeAmount", description="전일 대비 가격 차이 (원)")
    change_rate: float = Field(..., serialization_alias="changeRate", description="전일 대비 등락률 (%)")
    trend: str = Field(..., description="등락 추세 (up/down/flat)")
    volume: int = Field(..., description="거래량 (주)")
    trading_value: int = Field(..., serialization_alias="tradingValue", description="거래대금 (원)")
    market_cap: Optional[int] = Field(None, serialization_alias="marketCap", description="시가총액 (원)")
    is_favorite: bool = Field(False, serialization_alias="isFavorite", description="관심종목 여부")


class MarketQuoteListResponse(BaseModel):
    """시세 리스트 응답"""
    model_config = ConfigDict(populate_by_name=True)

    items: List[MarketQuoteItem]
    total: int = Field(..., description="전체 종목 수")
    page: int = Field(..., description="현재 페이지")
    page_size: int = Field(..., serialization_alias="pageSize", description="페이지 크기")
    has_next: bool = Field(..., serialization_alias="hasNext", description="다음 페이지 존재 여부")

File: SL-Front-End/src/lib/api/market-quote.ts
import { axiosInstance } from "../axios";

export interface MarketQuoteItem {
  rank: number;
  name: string;
  code: string;
  theme?: string | null;
  price: number;
  changeAmount: number;
  changeRate: number;
  trend: "up" | "down" | "flat";
  volume: number;
  tradingValue: number;
  marketCap: number | null;
  isFavorite: boolean;
}

export interface MarketQuoteListResponse {
  items: MarketQuoteItem[];
  total: number;
  page: number;
  pageSize: number;
  hasNext: boolean;
}

export type SortBy =
  | "volume"
  | "change_rate"
  | "trading_value"
  | "market_cap"
  | "name";
export type SortOrder = "asc" | "desc";

// 관심 종목 관련 타입
export interface FavoriteStockItem {
  stockCode: string;
  stockName: string;
  theme?: string | null;
  currentPrice: number | null;
  changeRate: number | null;
  previousClose: number | null;
  volume: number | null;
  tradingValue: number | null;
  marketCap: number | null;
  createdAt: string;
}

export interface FavoriteStockListResponse {
  items: FavoriteStockItem[];
  total: number;
}

// 최근 본 종목 관련 타입
export interface RecentStockItem {
  stockCode: string;
  stockName: string;
  currentPrice: number | null;
  changeRate: number | null;
  previousClose: number | null;
  volume: number | null;
  tradingValue: number | null;
  marketCap: number | null;
  viewedAt: string;
}

export interface RecentStockListResponse {
  items: RecentStockItem[];
  total: number;
}

export const marketQuoteApi = {
  /**
   * 시세 페이지 데이터 조회
   */
  getMarketQuotes: async (params: {
    sortBy?: SortBy;
    sortOrder?: SortOrder;
    page?: number;
    pageSize?: number;
    userId?: string;
    search?: string;
  }): Promise<MarketQuoteListResponse> => {
    const { data } = await axiosInstance.get<MarketQuoteListResponse>(
      "/market/quotes",
      {
        params: {
          sort_by: params.sortBy || "market_cap",
          sort_order: params.sortOrder || "desc",
          page: params.page || 1,
          page_size: params.pageSize || 50,
          user_id: params.userId,
          search: params.search,
        },
      },
    );
    return data;
  },

  /**
   * 관심 종목 목록 조회
   */
  getFavorites: async (): Promise<FavoriteStockListResponse> => {
    const { data } =
      await axiosInstance.get<FavoriteStockListResponse>("/market/favorites");
    return data;
  },

  /**
   * 관심 종목 추가
   */
  addFavorite: async (stockCode: string): Promise<void> => {
    await axiosInstance.post("/market/favorites", { stock_code: stockCode });
  },

  /**
   * 관심 종목 삭제
   */
  removeFavorite: async (stockCode: string): Promise<void> => {
    await axiosInstance.delete(`/market/favorites/${stockCode}`);
  },

  /**
   * 최근 본 종목 목록 조회
   */
  getRecentViewed: async (): Promise<RecentStockListResponse> => {
    const { data } = await axiosInstance.get<RecentStockListResponse>(
      "/market/recent-viewed",
    );
    return data;
  },

  /**
   * 최근 본 종목 추가 (자동 기록)
   */
  addRecentViewed: async (stockCode: string): Promise<void> => {
    await axiosInstance.post(`/market/recent-viewed/${stockCode}`);
  },

  /**
   * 최근 본 종목 삭제 (수동 삭제)
   */
  removeRecentViewed: async (stockCode: string): Promise<void> => {
    await axiosInstance.delete(`/market/recent-viewed/${stockCode}`);
  },
};
